### ISSUE TITLE
Should `extendr-api` be thread safe?

### ISSUE THREAD

**Issue by CGMossa** [extendr/extendr#659]
Dear all, I think it is time, that we discussed this topic at length. 

Should `extendr-api` be thread safe? Meaning, should it be possible to use `Robj` in threads?

1. ExtendR should not support multithreaded use, as R is single-threaded anyway, and in order to
use parallelism, we have to stick to accessing the R-API in a single-threaded way, which is currently accomplished
via `single_threaded`-guard everywhere that functions from `libR-sys` are invoked, that seemingly allocates or mutates.
2. Make `extendr-api` completely thread-safe.
3. Choose option (1), but provide tools for people to accomplish multithreaded use by hand.

There are many reasons that we are discussing this. `single-threaded` was developed due to issues with testing, but
it does provide guarantees that translate to multithreaded code.
The maintenance burden is growing ever more, as we constantly have issue with multiple threads, etc.
The current implementation of `single-threaded` has a sleep in it, thus we are essentially slowing down even 1-thread usage of ExtendR.

_My personal opinion_ on this matter is twofold: Currently, we have `single-threaded` in our API, and thus I believe we should
at least make sure that this works first. I would like to consider the option of not promising multithreaded in extendr, but give tools
for users to write these themselves, mainly a functioning `single_threaded`-guard, and a wrapper for `Robj` that is `Send` (and maybe `Sync`?). As of right now, I believe that we're fairly close to the first goal. The second goal is being discussed in this thread.

Therefore, no matter what we decide here, we have to discuss, review and merge the following:

## Previous discussions

This issue is mainly raised due to https://github.com/extendr/extendr/issues/583#issuecomment-1752184772,
however after carefully trying to deal with this, I do clearly see the ever growing maintenance burden due to this.
It isn't fun. It isn't possible to spend `n` hours and solve any given issue. It is always many people required to even
put a dent in the issues that arise from this.

Another commentary from me is here: https://github.com/extendr/extendr/issues/583#issuecomment-1752697578.
This references having embedded R in extendr as well, that is also a contentious issue, that need similar discussions like this one.
But @yutannihilation have raised this in #583. 

### Tracking issue for `extendr` with multithreaded

- [ ] #635 that is now superseded by #658. I have removed the "thread sleep" from the implementation,
and I have had an external Rust expert review the PR (see #635).
- [ ] Make a PR where `Robj` implements `Send`, as currently it does **not**, but we have already made that
promise that it should if it is `Send`.
- [ ] Discuss what `Sync` means for R and for extendr, as I don't have a clear view of this atm.
- [ ] Add multithreaded tests to extendr

### Tracking issue for removing thread-safety from `extendr`.

- [ ] PR that removes all `single_threaded` use.
- [ ] Remove `Send` from `Robj`. 
- [ ] Provide a `SendRobj` that wraps `Robj`, and it must somehow check if it is in a `single_threaded`, and complain if it isn't.

**Comment by JosiahParry** [comment]
Here's my user take on parallelism: 

One of hte main reasons to use Rust is because of its "fearless concurrency." As a novice user of Rust, I find this to be one of the major reasons to use it! Rayon and tokio are game changing. I personally use both rayon and tokio in tandem with extendr but it requires a lot of hoop jumping. 

There are three ways that I would use parallelism with extendr. 

1. Using rust native types as a parallel iterator, collecting into a native Rust type, and converting into an R type
    - This is supported today
2. Using a rust native type as a parallel iterator, collecting into an Robj / R type
    - this is _not_ supported today.
3. Using R native types as a parallel iterator and collecting into an Robj / Robj
    - this is not supported today.
    - Example use case: I have a `List` of structs (`Geom`s in `{rsgeo}`) which I iterate over in parallel and call a method on it, for example `x.unsigned_area()` and collect it into `Doubles`
   
Today I introduce parallelism by taking a `List` of structs, cloning them into a Rust owned vector, using Rayon, then collecting into a Vector of non-extendr types. Then in another iteration convert them into an extendr type. There becomes 2 extra iterations over my vector in that case to use parallelism. In some cases that has caused parallel iteration to be slower and less efficient than single threaded nature.  Example [here](https://github.com/JosiahParry/rsgeo/blob/7623f3015a4a17f59ad9bab6f603f09547fd0063/src/rust/src/length.rs#L71).

I would like to be able to do at least 2. I envision this as having some rust type that I can introduce as a parallel iterator. A real example is [here](https://github.com/JosiahParry/arcpbf/blob/23647528f693397a1b6078e1520aacea0097c567/src/rust/arcpbf/src/lib.rs#L49C14-L49C14). Ideally I could take a `Vec<Vec<u8>>` and process this in parallel using rayon. In each iteration I create an `Robj` that I would like to `collect::<Vec<Robj>>()`. After processing, I would turn this into a `List::from_values(res_vec)`. However, this is not possible because even if the `Robj` has not seen R yet, it cannot be sent between threads. I'd like something like `SendRobj` or `UnsafeRobj` or something that I can create in parallel and collect on a main thread so I can at least _create_ `Robj`s in parallel. 

I understand the difficulty of #3 and don't foresee that being feasible right away. Nor would I need to do that. But I find 2 to be frustrating becasue the data (at least from my perspective) as not even seen R yet it just has an Robj type.
